src/khoj/processor/conversation/gpt4all/chat_model.py

```diff
@@ -0,0 +1,137 @@
+from typing import Union, List
+from datetime import datetime
+import sys
+import logging
+from threading import Thread
+
+from langchain.schema import ChatMessage
+
+from gpt4all import GPT4All
+
+
+from khoj.processor.conversation.utils import ThreadedGenerator, generate_chatml_messages_with_context
+from khoj.processor.conversation import prompts
+from khoj.utils.constants import empty_escape_sequences
+
+logger = logging.getLogger(__name__)
+
+
+def extract_questions_falcon(
+    text: str,
+    model: str = "ggml-model-gpt4all-falcon-q4_0.bin",
+    loaded_model: Union[GPT4All, None] = None,
+    conversation_log={},
+    use_history: bool = False,
+    run_extraction: bool = False,
+):
+    """
+    Infer search queries to retrieve relevant notes to answer user query
+    """
+    all_questions = text.split("? ")
+    all_questions = [q + "?" for q in all_questions[:-1]] + [all_questions[-1]]
+    if not run_extraction:
+        return all_questions
+
+    gpt4all_model = loaded_model or GPT4All(model)
+
+    # Extract Past User Message and Inferred Questions from Conversation Log
+    chat_history = ""
+
+    if use_history:
+        chat_history = "".join(
+            [
+                f'Q: {chat["intent"]["query"]}\n\n{chat["intent"].get("inferred-queries") or list([chat["intent"]["query"]])}\n\nA: {chat["message"]}\n\n'
+                for chat in conversation_log.get("chat", [])[-4:]
+                if chat["by"] == "khoj"
+            ]
+        )
+
+    prompt = prompts.extract_questions_falcon.format(
+        chat_history=chat_history,
+        text=text,
+    )
+    message = prompts.general_conversation_falcon.format(query=prompt)
+    response = gpt4all_model.generate(message, max_tokens=200, top_k=2)
+
+    # Extract, Clean Message from GPT's Response
+    try:
+        questions = (
+            str(response)
+            .strip(empty_escape_sequences)
+            .replace("['", '["')
+            .replace("']", '"]')
+            .replace("', '", '", "')
+            .replace('["', "")
+            .replace('"]', "")
+            .split('", "')
+        )
+    except:
+        logger.warning(f"Falcon returned invalid JSON. Falling back to using user message as search query.\n{response}")
+        return all_questions
+    logger.debug(f"Extracted Questions by Falcon: {questions}")
+    questions.extend(all_questions)
+    return questions
+
+
+def converse_falcon(
+    references,
+    user_query,
+    conversation_log={},
+    model: str = "ggml-model-gpt4all-falcon-q4_0.bin",
+    loaded_model: Union[GPT4All, None] = None,
+    completion_func=None,
+) -> ThreadedGenerator:
+    """
+    Converse with user using OpenAI's ChatGPT
```

Comment:
The function comment should be fixed
```suggestion
    Converse with user using Falcon
```